### PR TITLE
resource_control: take global resource consumption into consideration

### LINKED ISSUE #14604: Resource control is not ideal when the request load from different resource group is unbalanced #6335
## Development Task

Consider the case that there are 3 stores, and the request load from one resource group is a hotspot workload. Then all requests from the resource group are handled on one store whereas the requests from the other resource group are distributed evenly on all stores.

As resource control on TiKV only has the local resource consumption, the local QPS ratio may meet the ratio of resource quota of different resource groups but the overall QPS ratio may not meet with skew load. To solve that, the request to one store needs to carry the resource consumption on other stores to make the overall consumption considered.

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: Close #14604

What's Changed:

```commit-message
resource control takes global resource consumption into consideration
```

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

Under an unbalanced leader distribution, this patch can control the QPS quota of two resource groups nearly 1:1
img_v2_1f407564-81bf-4718-a8a3-6805abe71aag

### Release note 

```release-note
resource control takes global resource consumption into consideration to improve scheduling effect
```

### CODE REVIEW

File: components/resource_control/src/resource_group.rs [hunk 1]
```diff
@@ -143,6 +143,25 @@ impl ResourceGroupManager {
             controller.update_min_virtual_time();
         }
     }
+
+    pub fn consume_penalty(&self, ctx: &ResourceControlContext) {
```
Comment (Connor1996) [inline components/resource_control/src/resource_group.rs:147]: FYI, the core logic is here, other places are mostly boilerplate

File: src/server/service/kv.rs [hunk 2]
```diff
@@ -177,9 +183,12 @@ macro_rules! handle_request {
             let begin_instant = Instant::now();
 
             let source = req.mut_context().take_request_source();
-            let resource_group_name = req.get_context().get_resource_group_name();
+            let resource_control_ctx = req.get_context().get_resource_control_context();
```
Comment (glorv) [inline src/server/service/kv.rs:186]: Please move this code into a function, it repeats several times.

**Review by glorv — APPROVED** [review]
LGTM

File: src/server/service/kv.rs (hunk 2 above)
Comment (Connor1996) [reply src/server/service/kv.rs:186]: can't because the reqs are of differrent type

**Review by nolouch — APPROVED** [review]
lgtm

File: src/server/service/kv.rs (hunk 2 above)
Comment (glorv) [reply src/server/service/kv.rs:186]: But you can just pass resource_control_ctx as the parameter

**Review by tabokie — APPROVED** [review]

**Pull request merged**